File: app/src/main/java/de/chrisward/theyworkforyou/view/LordListFragment.java
package de.chrisward.theyworkforyou.view;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.util.List;

import de.chrisward.theyworkforyou.R;
import de.chrisward.theyworkforyou.model.Lord;
import de.chrisward.theyworkforyou.view.adapter.LordAdapter;
import de.chrisward.theyworkforyou.viewmodel.LordViewModel;

public class LordListFragment extends Fragment implements LordAdapter.OnLordItemListener {
    private RecyclerView lordRecyclerView;
    private List<Lord> lords;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setRetainInstance(true);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.person_list_fragment, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        if (getActivity() == null) {
            return;
        }

        TextView personListTitle = getActivity().findViewById(R.id.personListTitle);
        personListTitle.setText(R.string.all_lords);

        lordRecyclerView = getActivity().findViewById(R.id.personRecyclerView);
        lordRecyclerView.setHasFixedSize(true);
        lordRecyclerView.addItemDecoration(new DividerItemDecoration(getActivity(), LinearLayoutManager.VERTICAL));
        lordRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));

        LordViewModel lordViewModel = ViewModelProviders.of(this).get(LordViewModel.class);

        lordViewModel.lordList.observe(this, lords -> {
            this.lords = lords;
            if (getAdapter() != null) {
                getAdapter().replaceAll(lords);
                return;
            }

            lordRecyclerView.setAdapter(
                    new LordAdapter(lords, getLayoutInflater(), this)
            );
        });

        if (savedInstanceState == null) {
            lordViewModel.refreshLords(this);
        }
    }

    private LordAdapter getAdapter() {
        return ((LordAdapter) lordRecyclerView.getAdapter());
    }

    @Override
    public void onLordItemClick(int position) {
        //TODO Handle click.
    }
}

File: app/src/main/java/de/chrisward/theyworkforyou/view/LordListFragment.kt
package de.chrisward.theyworkforyou.view

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.fragment.app.Fragment
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import dagger.android.support.AndroidSupportInjection

import de.chrisward.theyworkforyou.R
import de.chrisward.theyworkforyou.model.Lord
import de.chrisward.theyworkforyou.view.adapter.LordAdapter
import de.chrisward.theyworkforyou.viewmodel.LordViewModel
import javax.inject.Inject

class LordListFragment : Fragment(), LordAdapter.OnLordItemListener {
    private var lordRecyclerView: RecyclerView? = null
    private var lords: List<Lord> = ArrayList()

    private val adapter: LordAdapter?
        get() = lordRecyclerView!!.adapter as LordAdapter?

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        retainInstance = true
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.person_list_fragment, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        AndroidSupportInjection.inject(this)

        val personListTitle = activity!!.findViewById<TextView>(R.id.personListTitle)
        personListTitle.setText(R.string.all_lords)

        lordRecyclerView = activity!!.findViewById(R.id.personRecyclerView)
        lordRecyclerView!!.apply {
            layoutManager = LinearLayoutManager(activity)
            setHasFixedSize(true)
            addItemDecoration(DividerItemDecoration(activity, LinearLayoutManager.VERTICAL))
        }

        val lordViewModel = ViewModelProviders.of(this, viewModelFactory).get(LordViewModel::class.java)

        lordViewModel.lordList.observe(this, Observer<List<Lord>> { lords  ->
            this.lords = lords!!
            if (adapter != null) {
                adapter!!.replaceAll(lords)
            } else {
                lordRecyclerView!!.adapter = LordAdapter(lords, layoutInflater, this)
            }
        })

        if (savedInstanceState == null) {
            lordViewModel.refreshLords(this)
        }
    }

    override fun onLordItemClick(position: Int) {
        //TODO Handle click.
    }
}
